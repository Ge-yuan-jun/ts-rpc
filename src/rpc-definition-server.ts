import glob from 'glob'
import { Project, Node, StructureKind, ClassDeclaration, ClassDeclarationStructure, FunctionDeclaration, MethodDeclaration, SourceFile, InterfaceDeclaration, SyntaxKind, TypeReferenceNode, TypeAliasDeclaration } from 'ts-morph'
import path from 'path'

export async function startRPCDefinitionServer (filePaths: string): Promise<() => string> {
  const files = glob.sync(filePaths)
  // console.log(3333, files)
  const prj = new Project({ compilerOptions: { declaration: true } })
  const rpjSfs = files.map(file => {
    return prj.addSourceFileAtPath(file)
  })
  const serverSf = prj.addSourceFileAtPath(path.resolve(__dirname, './server.ts'))
  const rpcServiceDef = serverSf.getFunction('RPCService')
  const rpcMethodDef = serverSf.getFunction('RPCMethod')
  if (rpcServiceDef == null || rpcMethodDef == null) throw Error('Cannot find RPCService or RPCMethod definition')

  // console.log(111, rpcMethodDef)
  const relateClasses = rpcServiceDef?.findReferencesAsNodes()
    // symbol <- callexpress <- decorator <- class
    .map(ref => ref.getParent()?.getParent()?.getParent())
    .filter(n => Node.isClassDeclaration(n)) as ClassDeclaration[]

  const simpleAST = relateClasses.map(n => ({
    kind: StructureKind.Class,
    name: n.getName(),
    methods: findRPCMethods(n, rpcMethodDef)
  })) as ClassDeclarationStructure[]

  // console.log(111, relateClasses)
  const genSf = prj.createSourceFile(path.resolve(__dirname, '../dist/__gen__.ts'), {
    statements: simpleAST
  }, { overwrite: true })
  rpjSfs.forEach((sf) => {
    const uiStruct = sf.getInterface('UserInfo')?.getStructure()
    if (uiStruct != null) genSf.addInterface(uiStruct)
  })
  // sf.addClass({ name: 'Abc' }).getStructure()
  // await sf.save()
  console.log(222, genSf.getEmitOutput({ emitOnlyDtsFiles: true }).getOutputFiles().map(file => file.getText()))
  return () => ''
}

function findRPCMethods (service: ClassDeclaration, rpcMethodDef: FunctionDeclaration): Array<ReturnType<MethodDeclaration['getStructure']>> {
  return service.getMethods()
    .filter(m => m.getDecorators().some(
      d => d.getNameNode()
        .getDefinitionNodes()
        .some(n => n === rpcMethodDef)
    ))
    .map(m => m.getStructure())
}

export function collectMethodTypeDeps (method: MethodDeclaration, sfs: SourceFile[]): Array<InterfaceDeclaration | TypeAliasDeclaration> {
  const params = method.getParameters()
  // console.log(2222, method.getSignature().getReturnType())
  // console.log(1111, params.map(p => p.getStructure()))
  const paramsDeclaration = params.map(
    p => p.getTypeNode()
      ?.getChildrenOfKind(SyntaxKind.Identifier)
  )
    .flat()
    .concat(
      method.getReturnTypeNode()
        ?.getChildrenOfKind(SyntaxKind.Identifier)
    )
    .map(i => i?.getDefinitionNodes())
    .flat()
    // .map(d => (d as InterfaceDeclaration))
    // .map(d => (d as InterfaceDeclaration).getStructure())
  // console.log(1111, paramsIdentifier)
  // console.log(1111, JSON.stringify(paramsIdentifier, null, 2))
  return paramsDeclaration as Array<InterfaceDeclaration | TypeAliasDeclaration>
}

type ITDeclaration = InterfaceDeclaration | TypeAliasDeclaration
export function collectTypeDeps (t: Node): ITDeclaration[] {
  const typeRefDeclarationSet = new Set<ITDeclaration>()
  function accessTRF (n: Node): void {
    n.forEachChild(c => {
      if (c instanceof TypeReferenceNode) {
        c.getChildrenOfKind(SyntaxKind.Identifier)
          .map(i => i.getSymbol())
          .flat()
          .map(s => s?.getDeclarations())
          .flat()
          .forEach(d => {
            if (
              d != null &&
              (
                d instanceof InterfaceDeclaration ||
                d instanceof TypeAliasDeclaration
              )
            ) typeRefDeclarationSet.add(d)
          })
      }
      accessTRF(c)
    })
  }
  accessTRF(t)
  return Array.from(typeRefDeclarationSet)
}
