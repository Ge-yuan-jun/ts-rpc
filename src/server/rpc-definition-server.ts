import glob from 'glob'
import { Project, Node, ClassDeclaration, FunctionDeclaration, MethodDeclaration, InterfaceDeclaration, SyntaxKind, TypeReferenceNode, TypeAliasDeclaration, MethodSignature } from 'ts-morph'
import path from 'path'

export async function startRPCDefinitionServer (filePaths: string): Promise<() => string> {
  const files = glob.sync(filePaths)
  const prj = new Project({ compilerOptions: { declaration: true } })
  files.forEach(file => {
    return prj.addSourceFileAtPath(file)
  })
  const serverSf = prj.addSourceFileAtPath(path.resolve(__dirname, './index.ts'))
  const rpcServiceDef = serverSf.getFunction('RPCService')
  const rpcMethodDef = serverSf.getFunction('RPCMethod')
  if (rpcServiceDef == null || rpcMethodDef == null) throw Error('Cannot find RPCService or RPCMethod definition')

  const refedClasses = rpcServiceDef?.findReferencesAsNodes()
    // symbol <- callexpress <- decorator <- class
    .map(ref => ref.getParent()?.getParent()?.getParent())
    .filter(n => Node.isClassDeclaration(n)) as ClassDeclaration[]

  const genSf = prj.createSourceFile(path.resolve(__dirname, '_protocol-file-memory_.ts'), {
  }, { overwrite: true })

  refedClasses.forEach((c) => {
    const className = c.getName()
    if (className == null) throw Error('RPCService must be applied to a named class')
    // 将 class 转换为 interface，模拟 rpc 的 protocol 声明
    const inter = genSf.addInterface({ name: className })
    inter.addJsDocs(c.getJsDocs().map(doc => doc.getStructure()))
    const methods = findRPCMethods(c, rpcMethodDef)
    inter.addMethods(
      methods.map(m => (
        m.getSignature().getDeclaration() as MethodSignature
      ).getStructure())
    )

    methods.map(m => collectMethodTypeDeps(m))
      .flat()
      .forEach(it => {
        // 添加 method 依赖的类型，否则无法编译通过
        if (it instanceof InterfaceDeclaration) {
          genSf.insertInterface(0, it.getStructure())
        } else if (it instanceof TypeAliasDeclaration) {
          genSf.insertTypeAlias(0, it.getStructure())
        } else {
          // TODO: error
        }
      })
  })

  const protocolFileContent = genSf.getEmitOutput({ emitOnlyDtsFiles: true })
    .getOutputFiles().map(file => file.getText())
    .join('\n')
  return () => protocolFileContent
}

function findRPCMethods (service: ClassDeclaration, rpcMethodDef: FunctionDeclaration): MethodDeclaration[] {
  return service.getMethods()
    .filter(m => m.getDecorators().some(
      d => d.getNameNode()
        .getDefinitionNodes()
        .some(n => n === rpcMethodDef)
    ))
}

type ITDeclaration = InterfaceDeclaration | TypeAliasDeclaration
export function collectMethodTypeDeps (method: MethodDeclaration): ITDeclaration[] {
  return [...method.getParameters(), method.getReturnTypeNodeOrThrow()]
    .map(n => collectTypeDeps(n))
    .flat()
}

export function collectTypeDeps (t: Node): ITDeclaration[] {
  const typeRefDeclarationSet = new Set<ITDeclaration>()

  if (t instanceof TypeReferenceNode) {
    findITDeclaration(t)
  } else if (isITDeclaration(t)) {
    typeRefDeclarationSet.add(t)
  }

  // 深度优先遍历树，找到引用Type类型，然后找到 Declaration
  queryInTree(t)

  return Array.from(typeRefDeclarationSet)

  function queryInTree (n: Node): void {
    n.forEachChild(c => {
      if (c instanceof TypeReferenceNode) findITDeclaration(c)

      queryInTree(c)
    })
  }

  function isITDeclaration (n: Node): n is ITDeclaration {
    return n instanceof InterfaceDeclaration ||
      n instanceof TypeAliasDeclaration
  }

  function findITDeclaration (n: Node): void {
    n.getChildrenOfKind(SyntaxKind.Identifier)
      .map(i => i.getSymbol())
      .flat()
      .map(s => s?.getDeclarations())
      .flat()
      .forEach(d => {
        if (d != null && isITDeclaration(d)) typeRefDeclarationSet.add(d)
      })
  }
}
