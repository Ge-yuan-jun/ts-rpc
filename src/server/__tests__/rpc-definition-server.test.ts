import path from 'path'
import { MethodDeclaration, Project } from 'ts-morph'
import { collectMethodTypeDeps, collectTypeDeps, startRPCDefinitionServer } from '../rpc-definition-server'

test('startRPCDefinitionServer', () => {
  const server = startRPCDefinitionServer(path.resolve(__dirname, 'controller-example.ts'))

  const { dts, meta } = server()
  expect(dts).toMatchSnapshot()
  expect(meta).toEqual([{
    name: 'User',
    methods: [
      { name: 'getInfoById' },
      { name: 'getUnreadMsg' },
    ]
  }, {
    name: 'Foo',
    methods: [{ name: 'bar' }]
  }])
})

test('collectMethodTypeDeps', () => {
  const prj = new Project({ compilerOptions: { declaration: true } })
  const sf = prj.createSourceFile('test.ts', `
    interface A {}
    type B {}
    interface Z {}

    class Test {
      foo (a: A, b: B, c: string): Z {}
    }
  `)

  const deps = collectMethodTypeDeps(
    sf.getClass('Test')?.getMethod('foo') as MethodDeclaration
  )
  expect(deps.map(d => d.getName())).toEqual(['A', 'B', 'Z'])
})

test('collectTypeDeps', () => {
  const prj = new Project({ compilerOptions: { declaration: true } })
  const sf = prj.createSourceFile('test.ts', `
    interface A {
      b: B
    }
    interface B {
      a: A
    }

    interface X {
      a: A
      b: B
      ab: A | B
      s: string
      n: number
      f: () => any
    }
    type Y = A | B | null | undefined
  `)
  const iDeps = collectTypeDeps(sf.getInterfaceOrThrow('X'))
  expect(iDeps.map(i => i.getName())).toEqual(['X', 'A', 'B'])

  const tDeps = collectTypeDeps(sf.getTypeAliasOrThrow('Y'))
  expect(tDeps.map(t => t.getName())).toEqual(['Y', 'A', 'B'])
})
