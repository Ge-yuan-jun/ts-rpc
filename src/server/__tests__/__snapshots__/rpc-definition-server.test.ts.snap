// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`startRPCDefinitionServer 1`] = `
"interface UserInfo {
    name: string;
    age: number;
    avatar: string;
}
/**
 * service doc
 */
interface User {
    /**
     * method doc
     */
    getInfoById(id: string): UserInfo;
    getUnreadMsg(id: string): string[];
}
interface Foo {
    bar(): void;
}
"
`;
